overlord/fdestate/fdestate.go

```diff
@@ -187,100 +183,81 @@ func initializeState(st *state.State) error {
-	}
-
-	if fallbackCounterHandle != fallbackSaveCounterHandle {
-		return fmt.Errorf("the recover data and save keys do not use the same counter handle")
+		return fmt.Errorf("cannot get encrypted disks: %w", err)
 	}
 
+	s.PrimaryKeys = map[int]PrimaryKeyInfo{}
 	// Note that Parameters will be updated on first update
 	s.KeyslotRoles = map[string]KeyslotRoleInfo{
 		// TODO:FDEM: use a constant
 		"run": {
-			PrimaryKeyID:                   0,
-			TPM2PCRPolicyRevocationCounter: runCounterHandle,
+			PrimaryKeyID: 0,
 		},
 		// TODO:FDEM: use a constant
 		"run+recover": {
-			PrimaryKeyID:                   0,
-			TPM2PCRPolicyRevocationCounter: runCounterHandle,
+			PrimaryKeyID: 0,
 		},
 		// TODO:FDEM: use a constant
 		"recover": {
-			PrimaryKeyID:                   0,
-			TPM2PCRPolicyRevocationCounter: fallbackCounterHandle,
+			PrimaryKeyID: 0,
 		},
 	}
 
+	for _, disk := range disks {
+		statePrimaryKey, hasStatePrimaryKey := s.PrimaryKeys[0]
+		if !hasStatePrimaryKey {
+			digest, err := getPrimaryKeyDigest(disk.GetDevPath())
+			if err != nil {
+				if !errors.Is(err, secboot.ErrKernelKeyNotFound) {
+					return fmt.Errorf("cannot obtain primary key digest for data device %s: %w", disk.GetDevPath(), err)
+				}
+				logger.Noticef("cannot obtain primary key digest for data device %s: %v", disk.GetDevPath(), err)
+			} else {
+				s.PrimaryKeys[0] = PrimaryKeyInfo{Digest: digest}
+			}
+		} else {
+			sameDigest, err := statePrimaryKey.Digest.verifyPrimaryKeyDigest(disk.GetDevPath())
+			if err != nil {
+				if !errors.Is(err, secboot.ErrKernelKeyNotFound) {
+					return fmt.Errorf("cannot obtain primary key digest for data device %s: %w", disk.GetDevPath(), err)
+				}
+				logger.Noticef("cannot obtain primary key digest for data device %s: %v", disk.GetDevPath(), err)
+			} else if !sameDigest {
+				return fmt.Errorf("primary key for data and save partition are not the same")
+			}
+		}
+
+		legacyKeys := disk.GetLegacyKeys()
+		for _, keyName := range []string{"default", "default-fallback"} {
+			if keyName == "default" && disk.GetRole() == "system-save" {
+				continue
+			}
+			legacyKey, _ := legacyKeys[keyName]
+			handle, err := secbootGetPCRHandle(disk.GetDevPath(), keyName, legacyKey)
+			if err != nil {
+				return fmt.Errorf("cannot obtain counter handle for %s (default): %w", disk.GetDevPath(), err)
+			}
+			var profiles []string
+			switch keyName {
+			case "default":
+				profiles = []string{"run", "run+recover"}
+			case "default-fallback":
+				profiles = []string{"recover"}
+			}
+			for _, profile := range profiles {
+				role := s.KeyslotRoles[profile]
+				if s.KeyslotRoles[profile].TPM2PCRPolicyRevocationCounter == 0 {
+					role.TPM2PCRPolicyRevocationCounter = handle
+					s.KeyslotRoles[profile] = role
+				} else if role.TPM2PCRPolicyRevocationCounter != handle {
+					return fmt.Errorf("found multiple revocation count for run keys")
+				}
+			}
+		}
+	}
+
```

Comment:
should we log something or is it an error if at this stage PrimaryKeys is still empty?

Reply:
It was not an error before this PR, just a notice. I added a notice with "WARNING".